logaggregator/main_test.go

```diff
@@ -81,11 +86,191 @@ func (s *LogAggregatorTestSuite) TestAggregatorBuffersMessages(c *C) {
 		<-messageReceived // wait for messages to be received
 	}
 
-	msgs := s.agg.ReadLastN("app", 0)
-	c.Assert(msgs, HasLen, 2)
-	c.Assert(string(msgs[0].ProcID), Equals, "web.1")
-	c.Assert(string(msgs[1].ProcID), Equals, "web.2")
-	s.agg.Shutdown()
+	readAllMsgs := func(msgc <-chan *rfc5424.Message) []*rfc5424.Message {
+		timeout := time.After(2 * time.Second)
+		msgs := make([]*rfc5424.Message, 0)
+		for {
+			select {
+			case msg := <-msgc:
+				if msg == nil {
+					return msgs
+				}
+				msgs = append(msgs, msg)
+			case <-timeout:
+				c.Fatalf("timeout waiting for send on msgc")
+			}
+		}
+	}
+
+	tests := []struct {
+		lines           int
+		filters         []filter
+		expectedLen     int
+		expectedProcIDs []string
+	}{
+		{
+			lines:           -1,
+			filters:         nil,
+			expectedLen:     2,
+			expectedProcIDs: []string{"web.1", "web.2"},
+		},
+		{
+			lines: -1,
+			filters: []filter{
+				filterJobID{[]byte("1")},
+			},
+			expectedLen:     1,
+			expectedProcIDs: []string{"web.1"},
+		},
+		{
+			lines: 1,
+			filters: []filter{
+				filterJobID{[]byte("1")},
+			},
+			expectedLen:     1,
+			expectedProcIDs: []string{"web.1"},
+		},
+	}
+
+	for _, test := range tests {
+		c.Logf("test: %+v", test)
+		msgc := s.agg.ReadLastN("app", test.lines, test.filters, make(chan struct{}))
+		msgs := readAllMsgs(msgc)
+		c.Assert(msgs, HasLen, test.expectedLen)
+		for i, procID := range test.expectedProcIDs {
+			c.Assert(string(msgs[i].ProcID), Equals, procID)
+		}
+	}
+}
+
+func (s *LogAggregatorTestSuite) TestAggregatorReadLastNAndSubscribe(c *C) {
+	// set up testing hook:
+	messageReceived := make(chan struct{})
+	afterMessage = func() {
+		messageReceived <- struct{}{}
+	}
+	defer func() { afterMessage = nil }()
+
+	conn, err := net.Dial("tcp", s.agg.Addr)
+	c.Assert(err, IsNil)
+	defer conn.Close()
+
+	_, err = conn.Write([]byte(sampleLogLine1))
+	c.Assert(err, IsNil)
+	_, err = conn.Write([]byte(sampleLogLine2))
+	c.Assert(err, IsNil)
+
+	for i := 0; i < 2; i++ {
+		<-messageReceived // wait for messages to be received
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	msgc := s.agg.ReadLastNAndSubscribe("app", -1, nil, ctx.Done())
+	timeout := time.After(5 * time.Second)
+	for i := 0; i < 2; i++ {
+		select {
+		case msg := <-msgc:
+			c.Assert(msg, Not(IsNil))
+		case <-timeout:
+			c.Fatalf("timeout waiting for send on msgc")
+		}
+	}
+	select {
+	case msg := <-msgc:
+		c.Fatalf("unexpected message received: %+v", msg)
+	default:
+	}
+	cancel()
```

Comment:
I added the above ~18 lines to fix a panic I discovered when `s.agg.ReadLastNAndSubscribe()` was called with `n == -1`. This test is getting a bit out of hand now so it needs refactoring before a merge.